Ability to Customize HAProxy 2.x Error Page (#680)
openshift/enhancements: enhancements/ingress/customize-error-code-pages.md

---
title: customize-error-code-pages
authors:
  - "@miheer"
reviewers:
  - "@Miciah"
  - "@danehans"
  - "@knobunc"
approvers:
  - "@Miciah"
  - "@danehans"
  - "@knobunc"
creation-date: 2021-03-07 
last-updated: 2020-03-07
status: implementable 
see-also:
replaces:
superseded-by:
---
# Customize error code paged returned by the haproxy router
## Release Signoff Checklist
- [X] Enhancement is `implementable`
- [ ] Design details are appropriately documented from clear requirements
- [X] Test plan is defined
- [ ] Graduation criteria for dev preview, tech preview, GA
- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
## Summary
There is no supported and reasonable method to customize the HAProxy 2.x error page as deployed in OpenShift.
Customizing the HAProxy error page per our recommended procedure (documented here: https://access.redhat.com/solutions/3425711).

> does not work with OCP 4. It's also important that the right page as per user requirements is served at the right moment.

Reviewer: > Customizing the HAProxy error page per our recommended procedure (documented here: https://access.redhat.com/solutions/3425711). does not work with OCP 4

This solution was never intended to work on OCP 4, right? Also, I would suggest removing the spurious period after the parenthesis.

Author: I will reword it. Yes it is not meant for OCP 4

Example - 503 error page shall be returned if no pod is available. 

> But when the URI is wrong/does not exist, we should return 404 or a customized 404 page as per user requirements.

Reviewer: > But when the URI is wrong/does not exist, we should return 404 or a customized 404 page as per user requirements.

How does this enhancement accomplish this? Can multiple error pages (ie one HTML page for 503, and one HTML page for 404) be appended to the same error pages config map (and thus fed into HAProxy)? I would suggest elaborating on how a user specifics multiple error pages somewhere in this enhancement (it's not entirely clear to me).

Author: I think I will have add a array of strings for the api. I will update it.

## Motivation
In OCP 3.11 customizing error pages is possible via a [documented procedure](https://access.redhat.com/solutions/3425711)

> Customers resist customizing the HAProxy error page per our recommended procedure (documented here: https://access.redhat.com/solutions/3425711)

Reviewer: > Customers resist customizing the HAProxy error page per our recommended procedure (documented here: https://access.redhat.com/solutions/3425711)

Are you saying that customers in OCP 3.11 resist following the recommended procedure? If so, why? Or are you saying customers on OCP 4.x resist following this procedure for updating the HAProxy error pages (simply because it does not work)?

Author: Actually that's a different thing. Customers say 503 page is served even when the page is not found when it shall serve 404. The reason I think that happens is because we provide only 503 in our haproxy template

Author: I will make it clear

Also this procedure is not possible in OCP 4.
So while it is possible to define custom Error Pages, it's also important that the right page is served at the right moment. Currently 503 is 
returned if no pod is available and also when the URI has a wrong target. But when the URI is wrong/does not
exist, we should return 404 or something different as it otherwise may confuse end-customers.

### Goals
1. Enable the cluster administrator to specify custom error code pages for haproxy router to serve them at the right moment.

### Non-Goal
TBD

## Proposal
0. User creates a custom error code page configmap `http503page` in `openshift-config`
1. The IngressController API is extended by adding an optional
`HttpErrorCodePage` field with type string
```go

> //This used to sync the configmap having customizable error code page created by cluster admin in openshift-config namespace.

Reviewer: In general I would suggest using a single space after each comment indicator, ie `// <text>` instead of `//<text>`.

Author: OK

//to openshift-ingress which will be used by the router to serve appropriate error code pages.
//If this field is empty, the default values i.e default shipped error code pages will be served by the haproxy router.
HttpErrorCodePage string `json:"httpErrorCodePage,omitempty"`

```
Using the default, the router uses the default error pages used by haproxy.
The `HttpErrorCodePage` field enables the cluster administrator to
specify custom error code pages.
```yaml
apiVersion: operator.openshift.io/v1
kind: IngressController
metadata:
  name: default
  namespace: openshift-ingress-operator
spec:
  httpErrorCodePage: "http503page"  // http503page is the config map name with custom 

>                                     //503 error page created in openshift-config by cluster-admin. 

Reviewer 2: Yaml comments start with `#`.  But instead of comments in the yaml, please provide an example definition for the "http503page" configmap.

Author: OK

```
2. `errorfile` stanzas need to be added as per https://www.haproxy.com/blog/serve-dynamic-custom-error-pages-with-haproxy/
3. default error pages will be added in openshift/router for the haproxy.

> 4. A  new controller to sync the cluster admin created configmap having custom error pages from `openshift-config` 

Reviewer: Is syncing the config map from the `openshift-config` namespace to the `openshift-ingress` namespace necessary? Is there a reason we can't just use the config map in the original namespace directly?

Reviewer: Ah I see, thanks for the historical explanation. Thanks!

   to `openshift-ingress`.

> 5. changes to ingress controller's function `desiredRouterDeployment` present in `deployment.go`.

Reviewer: ```suggestion
5. changes to ingress controller's function `desiredRouterDeployment` present in `deployment.go`
```

Author: OK

   of openshift/cluster-ingress-operator to mount the custom error code page configmap into the router deployment.

### Validation
Omitting `spec.httpErrorCodePage`  specifies the default behavior i.e serves default error code pages.

> If user specifies an invalid value, then the IngressController ignores the annotation.

Reviewer: > If user specifies an invalid value, then the IngressController ignores the annotation.

What constitutes an invalid value? Which annotation are you referring to here?

Reviewer 2: In particular, what keys does the configmap need to have?  What is the format of the data values?  For example, does an error page need to include the HTTP header?  (Presumably it should.)  Does the page need to have an HTML document?  (I expect text/plain would work.)  If HTML, is it validated?  (I don't think we should validate HTML, but we should call out the design decision.)  Is there a max length?  Does encoding matter?

Author: Not yet decided on the max length of the page. Need to think on encoding.

### User Stories

#### As a cluster administrator, I expect to have custom error code pages getting served
1. The administrator creates a configmap named http503page in the openshift-config namespace.
2. The administrator patches the ingresscontroller to reference the http503page configmap by name.
3. The ingress operator copies the http503page configmap from the openshift-config namespace to the
   openshift-ingress namespace.
4. The ingress operator configures the router deployment with a volume and volume mount that mounts
   the custom error pages from the http503page configmap on top of the existing error pages in the 
   router pods.
   
### Risks and Mitigations

> No risk identified. This is a cosmetic change.

Reviewer: I think there are a couple risks worth mentioning.

1) What if a user supplies an incomplete or confusing error page(s)? Or what if a user provides an absolutely massive HTML file for a trivial 503? Could that effect performance somehow?
2) Are there any applications that use the default ingress controller that expect a 503 in non-503 cases? Does the administrator need to exercise caution when creating custom error pages for the default ingress controller? (I would hope not, but it might be worth double checking)

Author: If the user does not provide proper file name in configmap then the custom error page will not be served. The default shall be served in that case

Yeah admin will need to take some precaution. I will list them. May be we can do some validations at our end.

## Design Details

### Test Plan
Unit test cases will be added for the new controller which sync the configmap from openshift-config
to openshift-ingress.
Unit test cases will be added for the ingress controller where the router deployment get a volume mount

> of the custom error code page

Reviewer: I would suggestion adding some details about a potential e2e test for this enhancement.

Author: OK

---
title: customize-error-code-pages
authors:
  - "@miheer"
reviewers:
  - "@Miciah"
  - "@danehans"
  - "@knobunc"
approvers:
  - "@Miciah"
  - "@danehans"
  - "@knobunc"
creation-date: 2021-03-07 
last-updated: 2020-03-07
status: implementable 
see-also:
replaces:
superseded-by:
---
# Customize error code paged returned by the haproxy router
## Release Signoff Checklist
- [X] Enhancement is `implementable`
- [ ] Design details are appropriately documented from clear requirements
- [X] Test plan is defined
- [ ] Graduation criteria for dev preview, tech preview, GA
- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
## Summary

> There is no supported and reasonable method to customize the HAProxy 2.x error page in OCP 4.x
> It's also important that the right page as per user requirements is served at the right moment.
> Example - 503 error page shall be returned if no pod is available. 
> But when the URI is wrong/does not exist, we should return 404 or a customized 404 page as per user requirements.

Reviewer 2: ```suggestion
There is no supported method to customize an IngressController's error pages in OCP 4.
Users may want to customize an IngressController's error pages for branding or other reasons.
For example, users may want a custom HTTP 503 error page to be returned if no pod is available. 
When the requested URI does not exist, users may want an IngressController to return a custom 404 page.
```

> Using this feature, user can provide custom http pages for 503 and 404 for the haproxy router by adding these pages
> in a configmap using --file attribute.

Reviewer 2: What is the "--file attribute"?  I'm guessing the phrase can be deleted.  
```suggestion
Using this feature, users can provide custom HTTP pages for HTTP 503 and 404 responses using a configmap.
```

## Motivation
Users say 503 page is served even when the page is not found when it shall serve 404. 

> The reason that happens is because we provide only 503 in our haproxy template.

Reviewer 2: I think this a secondary motivation for this enhancement.  The primary motivation is that users want to customize the error pages that are returned, for example for branding purposes.  Please mention both motivations.

### Goals

> 1. Enable the cluster administrator to specify custom error code pages for haproxy router to serve them at the right moment.

Reviewer 2: From the motivation and summary, it seems you can also add this goal: "Return a distinct HTTP 404 error page when the requested URI does not exist, instead of returning the HTTP 503 error page."

I think you can delete the phrase "at the right moment".  If the phrase refers to returning 404s, then that should be a separate goal.  If the phrase means something else, please clarify.

### Non-Goal

> TBD

Reviewer 2: * Custom error pages for HTTP responses other than 503 and 404.
* Enabling users to configure custom error pages for individual routes.

## Motivation
Users say 503 page is served even when the page is not found when it shall serve 404. 
The reason that happens is because we provide only 503 in our haproxy template.

### Goals
1. Enable the cluster administrator to specify custom error code pages for haproxy router to serve them at the right moment.

### Non-Goal
TBD

## Proposal
0. User creates a custom error code page configmap `my-custom-error-code-pages` in `openshift-config`
1. The IngressController API is extended by adding an optional
`HttpErrorCodePages` field with type string
```go
	// httpErrorCodePages specifies a configmap with custom error pages.
	// The administrator must create this configmap in the openshift-config namespace.
	// This configmap should have keys in the format "error-page-<error code>.http",
	// where <error code> is an HTTP error code.
	// For example, "error-page-503.http" defines an error page for HTTP 503 responses.
	// Currently only error pages for 503 and 404 responses can be customized.
	// Each value in the configmap should be the full response, including HTTP headers.
	// If this field is empty, the ingress controller uses the default error pages.

> 	HttpErrorCodePages string `json:"httpErrorCodePages,omitempty"`

Reviewer 2: Shouldn't this be the configmap reference type?  
```suggestion
	HttpErrorCodePages configv1.ConfigMapNameReference `json:"httpErrorCodePages,omitempty"`
```

```

> Using the default, the router uses the default error pages used by haproxy.

Reviewer 2: ```suggestion
By default, an IngressController uses error pages built into the IngressController image.
```

Reviewer 2: This still needs to be fixed.

The `HttpErrorCodePage` field enables the cluster administrator to
specify custom error code pages.
```yaml
apiVersion: operator.openshift.io/v1
kind: IngressController
metadata:
  name: default
  namespace: openshift-ingress-operator
spec:
  httpErrorCodePage: "my-custom-error-code-pages"  // my-custom-error-code-pages is the config map name with custom 

>                                                    //http error pages, created in openshift-config by cluster-admin. 

Reviewer 2: Move the text after `//` into a paragraph following the code block that contains the yaml.

```

> 2. The configmap keys shall have the format error-page-`<error code>`.http. Right now only error-page-503.http and 
>    error-page-404.http will be available.

Reviewer 2: Add markup:
```suggestion
2. The configmap keys shall have the format `error-page-<error code>.http`. Right now only `error-page-503.http` and 
   `error-page-404.http` will be recognized.
```

```yaml
$ oc -n openshift-config create configmap  my-custom-error-code-pages\
--from-file=error-page-503.http\
--from-file=error-page-404.http

   $ oc get configmaps my-custom-error-code-pages -o yaml

apiVersion: v1
data:
   error-page-503.http: |-
     <your custom http response page contents>
   error-page-404.http: |

>      <your custom http response page contents>

Reviewer 2: When you get a chance, please fill out the example in more detail.  In particular, the example should show the HTTP headers to make it clear that the values aren't just HTML documents.

Reviewer 2: This still needs to be done.

kind: ConfigMap

```  

> 3. `errorfile` stanzas need to be added as per https://www.haproxy.com/blog/serve-dynamic-custom-error-pages-with-haproxy/

Reviewer 2: ```suggestion
3. In response to the user configuration, the IngressController will configure `errorfile` stanzas as needed in the IngressController's `haproxy.config` file per https://www.haproxy.com/blog/serve-dynamic-custom-error-pages-with-haproxy/
```
This is a little confusing because it seems like you doing two things at once:  You are both describing the procedure that the user would follow to use this feature, and you are also outlining the steps that the developer (you) will take to implement the feature.

> 4. default error pages will be added in openshift/router for the haproxy.

Reviewer 2: The image already has a default error page for 503s.  I guess you'll be adding a default error page for 404s, right?

Reviewer 2: This still needs to be addressed.  I think you mean to say the following:
```suggestion
4. A default error page for HTTP 404 responses will be added in openshift/router.
```

> 5. A  new controller to sync the cluster admin created configmap having custom error pages from `openshift-config` 
>    to `openshift-ingress`.

Reviewer 2: ```suggestion
5. A  new controller in the ingress operator will sync the configmap that the cluster admin created with custom error pages in the `openshift-config` namespace to the `openshift-ingress` namespace.
```

> 6. changes to ingress controller's function `desiredRouterDeployment` present in `deployment.go` to mount the configmap to the router.

Reviewer 2: ```suggestion
6. The ingress operator will configure the IngressController deployment to mount the configmap from the `openshift-ingress` namespace and use the error pages defined therein.
```

Reviewer 2: This still should be fixed.

### Validation
1. Omitting `spec.httpErrorCodePage`  specifies the default behavior i.e serves default error code pages.

> 2. If user specifies an invalid value, then the IngressController ignores the annotation.

Reviewer 2: ```suggestion
2. If the user specifies an invalid value, then the IngressController ignores the provided configmap.
```

> 3. If a user provides incorrect filename than specified format i.e error-page-<error code>.http in
> the keys of the configmap then it will be ignored and default page will be served.

Reviewer 2: ```suggestion
3. If the user provides an unrecognized key in the configmap (i.e., one that does not match the `error-page-<error code>.http` format), then the configmap will be ignored, and the default error page will be served.
```

Reviewer 2: This still needs to be fixed.

4. It is up to the user to ensure that the provided data is a valid HTTP response.

### User Stories

#### As a cluster administrator, I expect to have custom error code pages getting served

> 1. The administrator creates a configmap named http503page in the openshift-config namespace.

Reviewer 2: ```suggestion
1. The administrator creates a configmap named "http503page" in the `openshift-config` namespace.
```

Reviewer 2: This still isn't addressed.

2. The administrator patches the ingresscontroller to reference the http503page configmap by name.

> 3. The ingress operator copies the http503page configmap from the openshift-config namespace to the
>    openshift-ingress namespace.

Reviewer 2: ```suggestion
3. The ingress operator copies the http503page configmap from the `openshift-config` namespace to the
   `openshift-ingress` namespace.
```

Reviewer 2: This still isn't addressed.

4. The ingress operator configures the router deployment with a volume and volume mount that mounts
   the custom error pages from the http503page configmap on top of the existing error pages in the 
   router pods.
   
### Risks and Mitigations

> 1. If a user provides incorrect filename than specified format i.e error-page-<error code>.http in 
>    the keys of the configmap then it will be ignored and default page will be served. 

Reviewer 2: ```suggestion
1. If a user provides a key that is not of the format `error-page-<error code>.http` in 
   the configmap, then the configmap will be ignored, and default error pages will be served. 
```

2. It is up to the user to ensure that the provided data is a valid HTTP response.

## Design Details

### Test Plan

> Unit test cases will be added for the new controller which sync the configmap from openshift-config
> to openshift-ingress.

Reviewer 2: ```suggestion
Unit test cases will be added for the new controller which sync the configmap from `openshift-config`
to `openshift-ingress`.
```

Unit test cases will be added for the ingress controller where the router deployment get a volume mount
of the custom error code page

### Graduation Criteria
N/A.

### Upgrade / Downgrade Strategy
On upgrade, the default configuration does not perform error code page customization.  On
downgrade, the operator ignores the `spec.httpErrorCodePages` API field.

### Version Skew Strategy
N/A.

## Implementation History
Following are the most salient PRs in the feature's history:

> 1. https://github.com/openshift/api/pull/843  - Adds an field called `httpErrorCodePages` to the ingress controller API to get the custom error page configmap created 
>    in openshift-config namespace which is then synced to openshift-ingress namespace.

Reviewer 2: ```suggestion
1. https://github.com/openshift/api/pull/843  - Adds a field called `httpErrorCodePages` to the ingress controller API to get the custom error page configmap created 
   in the `openshift-config` namespace which is then synced to `openshift-ingress` namespace.
```

2. https://github.com/openshift/cluster-ingress-operator/pull/571 

>    a. Controller to sync configmaps from openshift config namespace to openshift-ingress namespace NE-535

Reviewer 2: ```suggestion
   a. Controller to sync configmaps from the `openshift-config` namespace to the `openshift-ingress` namespace ([NE-535](https://issues.redhat.com/browse/NE-535)).
```
Is NE-535 specifically for this part of the feature?

status: implementable 
see-also:
replaces:
superseded-by:
---
# Customize error code paged returned by the haproxy router
## Release Signoff Checklist
- [X] Enhancement is `implementable`
- [ ] Design details are appropriately documented from clear requirements
- [X] Test plan is defined
- [ ] Graduation criteria for dev preview, tech preview, GA
- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
## Summary
There is no supported method to customize an IngressController's error pages in OCP 4.
Users may want to customize an IngressController's error pages for branding or other reasons.
For example, users may want a custom HTTP 503 error page to be returned if no pod is available.
When the requested URI does not exist, users may want an IngressController to return a custom 404 page.
## Motivation
The primary motivation is that users want to customize the error pages that are returned, for example for branding purposes.
Secondary motivation is that users say 503 page is served even when the page is not found when it shall serve 404.
The reason that happens is because we provide only 503 in our haproxy template.

### Goals

> Return a distinct HTTP 404 error page when the requested URI does not exist, instead of returning the HTTP 503 error page.

Reviewer 2: Looks like you accidentally deleted the old goal: "Enable the cluster administrator to specify custom error code pages for haproxy router".

### Non-Goal
Custom error pages for HTTP responses other than 503 and 404.

> Enabling users to configure custom error pages for individual routes.

Reviewer 2: Use bullets.

	// The administrator must create this configmap in the openshift-config namespace.
	// This configmap should have keys in the format "error-page-<error code>.http",
	// where <error code> is an HTTP error code.
	// For example, "error-page-503.http" defines an error page for HTTP 503 responses.
	// Currently only error pages for 503 and 404 responses can be customized.
	// Each value in the configmap should be the full response, including HTTP headers.
	// If this field is empty, the ingress controller uses the default error pages.
	HttpErrorCodePages configv1.ConfigMapNameReference `json:"httpErrorCodePages,omitempty"`
```
Using the default, the router uses the default error pages used by haproxy.
The `HttpErrorCodePages` field enables the cluster administrator to
specify custom error code pages.
```yaml
apiVersion: operator.openshift.io/v1
kind: IngressController
metadata:
  name: default
  namespace: openshift-ingress-operator
spec:
  httpErrorCodePage: "my-custom-error-code-pages"  // my-custom-error-code-pages is the config map name with custom 
                                                   //http error pages, created in openshift-config by cluster-admin. 
```

> 2. The configmap keys shall have the format error-page-`<error code>`.http. Right now only error-page-503.http and
>    error-page-404.http will be available.

Reviewer 2: ```suggestion
2. The configmap keys shall have the format `error-page-<error code>.http`. Right now only `error-page-503.http` and
   `error-page-404.http` will be recognized.
```

```yaml

> $ oc -n openshift-config create configmap  my-custom-error-code-pages\
> --from-file=error-page-503.http\
> --from-file=error-page-404.http

Reviewer 2: ```suggestion
$ oc -n openshift-config create configmap  my-custom-error-code-pages \
--from-file=error-page-503.http \
--from-file=error-page-404.http
```
Or unwrap it:
```suggestion
$ oc -n openshift-config create configmap  my-custom-error-code-pages --from-file=error-page-503.http --from-file=error-page-404.http
```
